pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -542,3 +555,114 @@ func parseResourceGroupHashFromAnnotation(s *fleetv1beta1.ClusterResourceSnapsho
 	}
 	return v, nil
 }
+
+// TODO: handle multiple resourceSnapshots
+func (r *Reconciler) buildPlacementStatus(crp *fleetv1beta1.ClusterResourcePlacement, latestSchedulingPolicySnapshot *fleetv1beta1.ClusterSchedulingPolicySnapshot, latestResourceSnapshot *fleetv1beta1.ClusterResourceSnapshot) (*fleetv1beta1.ClusterResourcePlacementStatus, error) {
+	status := &fleetv1beta1.ClusterResourcePlacementStatus{
+		SelectedResources: make([]fleetv1beta1.ResourceIdentifier, 0, len(latestResourceSnapshot.Spec.SelectedResources)),
+		PlacementStatuses: make([]fleetv1beta1.ResourcePlacementStatus, 0, 20), // pre-allocate with a reasonable capacity
+		Conditions:        []metav1.Condition{},
+	}
+	// TODO query the sub resourceSnapshots by using latestResourceSnapshot and then update the status.
+	for i := range latestResourceSnapshot.Spec.SelectedResources {
+		obj, err := r.decodeResourceContent(&latestResourceSnapshot.Spec.SelectedResources[i])
+		if err != nil {
+			klog.ErrorS(err, "Failed to encode resourceContent of clusterResourceSnapshot", "clusterResourcePlacement", klog.KObj(crp), "clusterResourceSnapshot", klog.KObj(latestResourceSnapshot))
+			return nil, controller.NewUnexpectedBehaviorError(err)
+		}
+		ri := fleetv1beta1.ResourceIdentifier{
+			Group:     obj.GroupVersionKind().Group,
+			Version:   obj.GroupVersionKind().Version,
+			Kind:      obj.GroupVersionKind().Kind,
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		}
+		status.SelectedResources = append(status.SelectedResources, ri)
+	}
+	scheduledCondition := buildScheduledCondition(crp, latestSchedulingPolicySnapshot)
+	meta.SetStatusCondition(&status.Conditions, scheduledCondition)
+
+	if scheduledCondition.Status == metav1.ConditionUnknown {
+		appliedCondition := metav1.Condition{
+			Status:             metav1.ConditionUnknown,
+			Type:               string(fleetv1beta1.ClusterResourcePlacementAppliedConditionType),
+			Reason:             ApplyPendingReason,
+			Message:            "Scheduling has not completed",
+			ObservedGeneration: crp.Generation,
+		}
+		meta.SetStatusCondition(&status.Conditions, appliedCondition)
+		// skip populating detailed resourcePlacementStatus
+		return status, nil
+	}
+
+	// TODO collect work status and build appliedCondition & resourcePlacementStatus
+	status.PlacementStatuses = buildResourcePlacementStatus(crp, latestSchedulingPolicySnapshot)
+	return status, nil
+}
+
+func buildScheduledCondition(crp *fleetv1beta1.ClusterResourcePlacement, latestSchedulingPolicySnapshot *fleetv1beta1.ClusterSchedulingPolicySnapshot) metav1.Condition {
+	scheduledCondition := latestSchedulingPolicySnapshot.GetCondition(string(fleetv1beta1.PolicySnapshotScheduled))
+
+	if scheduledCondition == nil ||
+		scheduledCondition.ObservedGeneration < latestSchedulingPolicySnapshot.Generation ||
```

Comment:
The policy snapshot generation does not change much I am afraid (we do not update its spec); this comparison here might not be necessary at all.

Reply:
yeah, it's not needed, just for defensive coding.